File: clients/python/agentworth.py
"""AgentWorth Python client — a thin REST client over the HTTP ingress.

The ingress runs the SAME gate as everything else, so this client adds no
authority: a payment it submits is auto-executed inside a mandate, parked for
operator approval, or blocked. Standard-library only (no dependencies).

    from agentworth import AgentWorthClient

    os = AgentWorthClient("http://127.0.0.1:8787", token="...")
    result = os.pay(payee="tesco", payee_class="groceries", amount=8000,
                    rationale="the weekly grocery shop")
    print(result["outcome"])   # "settled" | "pending" | "blocked" | "failed"
"""

from __future__ import annotations

import json
import uuid
import urllib.error
import urllib.request
from typing import Any, Optional

__all__ = ["AgentWorthClient", "AgentWorthError"]


class AgentWorthError(Exception):
    """Raised on a transport error (not on a gate 'blocked' — that's a normal result)."""


class AgentWorthClient:
    def __init__(self, base_url: str = "http://127.0.0.1:8787", token: Optional[str] = None):
        self.base_url = base_url.rstrip("/")
        self.token = token

    def _request(self, method: str, path: str, body: Any = None, headers: Optional[dict] = None):
        data = json.dumps(body).encode() if body is not None else None
        h = {"content-type": "application/json"}
        if self.token:
            h["authorization"] = f"Bearer {self.token}"
        if headers:
            h.update(headers)
        req = urllib.request.Request(self.base_url + path, data=data, method=method, headers=h)
        try:
            with urllib.request.urlopen(req) as resp:
                payload = resp.read()
                return resp.status, (json.loads(payload) if payload else None)
        except urllib.error.HTTPError as e:
            payload = e.read()
            # 403/401/413/429 carry a JSON body too — return it, don't raise.
            return e.code, (json.loads(payload) if payload else None)
        except urllib.error.URLError as e:
            raise AgentWorthError(f"transport error: {e}") from e

    def pay(
        self,
        payee: str,
        payee_class: str,
        amount: int,
        currency: str = "GBP",
        rail: str = "card",
        rationale: str = "",
        idempotency_key: Optional[str] = None,
    ) -> dict:
        """Submit a payment intent. Returns the gate result
        {intentId, outcome, reasons, receiptId, verified}. `amount` is minor-units."""
        headers = {"idempotency-key": idempotency_key or str(uuid.uuid4())}
        _, body = self._request(
            "POST",
            "/payment-intent",
            {
                "payee": payee,
                "payeeClass": payee_class,
                "amount": amount,
                "currency": currency,
                "rail": rail,
                "rationale": rationale,
            },
            headers,
        )
        return body or {}

    def status(self) -> dict:
        """Kill-switch / circuit-breaker state."""
        return self._request("GET", "/status")[1] or {}

    def ready(self) -> dict:
        """Readiness probe."""
        return self._request("GET", "/ready")[1] or {}

    def openapi(self) -> dict:
        """The served OpenAPI 3.1 document."""
        return self._request("GET", "/openapi.json")[1] or {}

    def get_disclosure(self) -> dict:
        """Fetch this node's signed Verifiable Agency disclosure (public, no auth)."""
        return self._request("GET", "/.well-known/agent-disclosure")[1] or {}

    def verify_disclosure(self, disclosure: dict) -> dict:
        """Verifier-as-a-service: submit a signed disclosure, get a verdict
        {decision, tier, checks, reasons, cost}. Lets a heterogeneous counterparty
        verify a peer without implementing ed25519 itself. A "refuse" decision is a
        normal result, not an error."""
        return self._request("POST", "/verify-disclosure", disclosure)[1] or {}

File: clients/go/agentworth.go
// Package agentworth is a thin REST client over the AgentWorth HTTP ingress.
// The ingress runs the SAME gate as everything else, so this client adds no
// authority: a submitted payment is auto-executed inside a mandate, parked for
// operator approval, or blocked. Standard library only.
//
//	c := agentworth.New("http://127.0.0.1:8787", "token")
//	res, err := c.Pay(agentworth.PaymentIntent{
//	    Payee: "tesco", PayeeClass: "groceries", Amount: 8000,
//	    Currency: "GBP", Rail: "card", Rationale: "the weekly grocery shop",
//	}, "")
//	// res.Outcome is "settled" | "pending" | "blocked" | "failed"
package agentworth

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// Client talks to an AgentWorth ingress.
type Client struct {
	BaseURL string
	Token   string
	HTTP    *http.Client
}

// New builds a client. token may be empty for a loopback dev ingress.
func New(baseURL, token string) *Client {
	return &Client{
		BaseURL: strings.TrimRight(baseURL, "/"),
		Token:   token,
		HTTP:    http.DefaultClient,
	}
}

// PaymentIntent is a payment to submit. Amount is integer minor-units.
type PaymentIntent struct {
	Payee      string `json:"payee"`
	PayeeClass string `json:"payeeClass"`
	Amount     int64  `json:"amount"`
	Currency   string `json:"currency"`
	Rail       string `json:"rail"`
	Rationale  string `json:"rationale"`
}

// Result is the gate's verdict on a payment intent.
type Result struct {
	IntentID  string   `json:"intentId"`
	Outcome   string   `json:"outcome"` // settled | pending | blocked | failed
	Reasons   []string `json:"reasons"`
	ReceiptID *string  `json:"receiptId"`
	Verified  *bool    `json:"verified"`
}

func (c *Client) do(method, path string, body any, headers map[string]string) (int, []byte, error) {
	var rdr io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return 0, nil, err
		}
		rdr = bytes.NewReader(b)
	}
	req, err := http.NewRequest(method, c.BaseURL+path, rdr)
	if err != nil {
		return 0, nil, err
	}
	req.Header.Set("content-type", "application/json")
	if c.Token != "" {
		req.Header.Set("authorization", "Bearer "+c.Token)
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	return resp.StatusCode, data, err
}

// Pay submits a payment intent. idempotencyKey may be empty (one is generated).
// A "blocked" outcome is a normal result, not an error; err is non-nil only on a
// transport/decoding failure.
func (c *Client) Pay(intent PaymentIntent, idempotencyKey string) (*Result, error) {
	if idempotencyKey == "" {
		idempotencyKey = randomKey()
	}
	_, data, err := c.do("POST", "/payment-intent", intent, map[string]string{
		"idempotency-key": idempotencyKey,
	})
	if err != nil {
		return nil, err
	}
	var r Result
	if err := json.Unmarshal(data, &r); err != nil {
		return nil, fmt.Errorf("decode result: %w", err)
	}
	return &r, nil
}

// Verdict is the verifier-as-a-service result for a submitted disclosure.
type Verdict struct {
	Decision string          `json:"decision"` // transact | refuse
	Tier     string          `json:"tier"`     // cached | fresh
	Checks   map[string]bool `json:"checks"`
	Reasons  []string        `json:"reasons"`
}

// GetDisclosure fetches this node's signed Verifiable Agency disclosure (public).
func (c *Client) GetDisclosure() (json.RawMessage, error) {
	_, data, err := c.do("GET", "/.well-known/agent-disclosure", nil, nil)
	if err != nil {
		return nil, err
	}
	return json.RawMessage(data), nil
}

// VerifyDisclosure submits a signed disclosure to the verifier-as-a-service
// endpoint and returns the verdict. A "refuse" verdict is a normal result; err is
// non-nil only on a transport/decoding failure.
func (c *Client) VerifyDisclosure(disclosure json.RawMessage) (*Verdict, error) {
	_, data, err := c.do("POST", "/verify-disclosure", disclosure, nil)
	if err != nil {
		return nil, err
	}
	var v Verdict
	if err := json.Unmarshal(data, &v); err != nil {
		return nil, fmt.Errorf("decode verdict: %w", err)
	}
	return &v, nil
}

// Status returns the kill-switch / circuit-breaker state.
func (c *Client) Status() (map[string]any, error) {
	return c.getJSON("/status")
}

// Ready returns the readiness probe.
func (c *Client) Ready() (map[string]any, error) {
	return c.getJSON("/ready")
}

func (c *Client) getJSON(path string) (map[string]any, error) {
	_, data, err := c.do("GET", path, nil, nil)
	if err != nil {
		return nil, err
	}
	var m map[string]any
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	return m, nil
}

func randomKey() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}
